### DISCUSSION TITLE
A converter between yWriter and novelWriter

### DISCUSSION THREAD

**Discussion by peter88213** [vkbo/novelWriter#959] [General]
As I announced earlier, I wrote a Python script that converts projects between yWriter and novelWriter in both directions. Now the first alpha release can be downloaded.

For more information, see the [project homepage](https://peter88213.github.io/yw2nw) with description and download instructions.

Because of the different structures and features of yWriter and novelWriter, there are of course some limitations. All this is described in the instructions for use.  

Have fun with it!
Peter

**Comment by vkbo** [comment]
Very nice! I'll have to try it out. I haven't used yWriter in a long while, but as I've mentioned it's one of the original inspirations for making novelWriter. As a Linux user, yWriter is a bit inconvenient to use.

I should add this to a section in the README and docs for tools to use with novelWriter.

**Reply by peter88213** [reply]
When trying it out, please be sure to use the latest version, since it's still alpha state. However, I succeeded in converting my current 40.000 Words work in progress in both directions. 

As for running yWriter on Linux, WINE is the only way that works at the moment. First let winetricks install dotnet 3.5, then install the latest [yWriter5](http://spacejock.com/yWriter5_Download.html) release (.exe will work).

**Comment by vkbo** [comment]
Just a heads up that I will be changing the file format a bit in the next release. The updated project XML file looks like this: [nwProject.nwx](https://github.com/vkbo/novelWriter/blob/045a595537a6390639a56a609a667d92271aa4cf/sample/nwProject.nwx).

The changes are explained in #993.

They are quite simple. Since the `<item>` node had a lot of child nodes, I converted nearly all of them into attributes of the `<item>` and `<name>` nodes, and added a new `<meta>` node. If you look at the diff of the NWItem class that does the reading and writing, you'll see that it is a fairly simple change: [novelwriter/core/item.py](https://github.com/vkbo/novelWriter/commit/045a595537a6390639a56a609a667d92271aa4cf#diff-d9b0f36b7d63fb22b1d83ca4167b03d9f5f0f7ca7beed9a96634cbdfdef978feL142).

For the converter, I bet you can drop the `<meta>` node entirely as it doesn't contain any information I suppose you can transfer between the applications anyway. The `type`, `class`, and `layout` settings are now attributes of `<item>`. The `<name>` node is the same as before, but the `exported` flag and the the user-defined `status` value are attributes of `<name>`. Most values can be omitted, and will have a default setting.

The format change has been merged to the main branch now, but is not included in the 1.6 release I just made. It will not be out in the wild until 1.7. The file format has changed version from 1.3 to 1.4.

The reason for making this change is that the new format is a lot more compact and I plan to make changes to novelWriter to support multi-novel projects, which will mean the number of item nodes may grow a lot in large projects. I haven't changed how items are stored since the first versions of novelWriter, so the list of nodes has been growing quite large because of the verbose format.

**Reply by peter88213** [reply]
Thanks for the information, Veronica.
Since I abandoned my novelWriter fork, I don't have the most recent sources at the moment.
But after comparing your new sample project with the old one, the necessary change to my NwItem class was accomplished quite quickly.

Old code example for reading metadata:
```
        if node.find('status') is not None:
            self.nwStatus = node.find('status').text

        if node.find('exported') is not None:
            self.nwExported = node.find('exported').text
```
New code example:
```
        if node.find('meta') is not None:
            nwMeta = node.find('meta')
            self.nwStatus = nwMeta.attrib.get('status')
            self.nwExported = nwMeta.attrib.get('exported')
```
By the way, some "meta" data can certainly be exchanged between novelWriter and yWriter; I put some effort into mapping editing status for example. And the "exported" property lets yWriter make the difference between "used" and "unused" chapters and scenes.

At the moment I am preparing a completely refactored and thoroughly documented new version of my _PyWriter_ library. As soon as I'm done with it, I'll take a look at the new novelWriter version and test the changes to the converter.

Cheers,
Peter

**Reply by vkbo** [reply]
Great! I'm glad it was that simple to change the parsing/conversion.

By the way, `exported` should be a flag on the `<name>` node, not `<meta>`. It is a user setting. There is an attribute called `expanded` in `<meta>` though, which is just the last collapse/expand state of the element in the project tree on the GUI.

I put all structural info on the `<item>` node, all meta data on `<meta>` and all user-defined values on the `<name>` node. I though that made the most sense.

**Reply by peter88213** [reply]
Oops! That's right. Never mind, so the code for reading is like this:
```
        if node.find('name') is not None:
            nameNode = node.find('name')
            self.nwName = nameNode.text
            self.nwStatus = nameNode.attrib.get('status')
            self.nwExported = nameNode.attrib.get('exported')   
```
and for writing  like that:
```
        nameNode = ET.SubElement(node, 'name')        
        if self.nwName is not None:
            nameNode.text = self.nwName
        if self.nwStatus is not None:
            nameNode.set('status', self.nwStatus)
        if self.nwExported is not None:
            nameNode.set('exported', self.nwExported)
``` 
So you're right, the meta node is not necessary for converting from novelWriter to yWriter. 
However, scene word count and character count are available in yWriter and thus exported to novelWriter.

**Reply by vkbo** [reply]
> So you're right, the meta node is not necessary for converting from novelWriter to yWriter. However, scene word count and character count are available in yWriter and thus exported to novelWriter.

I added the automatic build of the index if it is empty (actually if any containing document is missing in it). It is included in today's release. Due to this, all word counts are re-calculated when you open the project the first time anyway. But there's no harm in exporting them still!

**Reply by peter88213** [reply]
In my case I have defined the scene content as a property; the setter calculates the word count on each change. With Python it looks like this:
```
    @property
    def sceneContent(self):
        return self._sceneContent

    @sceneContent.setter
    def sceneContent(self, text):
        """Set sceneContent updating word count and letter count."""
        self._sceneContent = text
        text = re.sub('\[.+?\]|\.|\,| -', '', self._sceneContent)
        # Remove yWriter raw markup for word count
        wordList = text.split()
        self.wordCount = len(wordList)
```
But may not be enough to have a living word counter in the editor.
As for novelWriter saving word counts, etc., I had suspected it was mainly for file previewing.

**Reply by vkbo** [reply]
The word counter in novelWriter is not quite that simple. See the code:
https://github.com/vkbo/novelWriter/blob/011291f225d7032f375b4473a43a0749e27cf92d/novelwriter/core/index.py#L823-L897

In any case, the information saved in the project XML is just to be able to quickly populate the tree without having to run the counter on the files. So the value is just a cached number. With the meta data there, no text files are loaded at all during start-up. This makes start-up a lot faster for large projects. The numbers are only duplicated into the project XML because the index isn't guaranteed to exist on start-up. When a document is being actively edited, the counts are re-computed every 5 seconds by a background thread.

**Reply by peter88213** [reply]
Basically, we use the same word counting algorithm: `len(text.split())` Because yWriter separates metadata from content from the start, it's a bit easier overall. Nevertheless, the word count of my converters almost never matches that of yWriter itself, just as each application provides different numbers (especially at NaNoWriMo, this is a constant topic of discussion). 
If you want to be very precise here, you have to agree, for example, whether abbreviations like "It's" or hyphenated composites in different languages should count as one or two words. Then you would also have to look for true and false apostrophes and so on. Actually unimportant, I think, but with 50,000+ words the differences do hurt one or another.

**Reply by vkbo** [reply]
Yeah, my code mostly just filters out formatting syntax. The rule I use for splitting words is indeed just whitespace, but I do also include dashes as word separators. I just replace them in the text before running the split, so both American style long dash pauses with no spaces, and short dashes used to define ranges and such, count as two words. Hyphenations count as single words.

The problem with getting too nit-picky about word boundary definitions is of course that it depends very much on the language. In Norwegian, as with German, we have a tendency to make words by combining multiple other words. Generally without hyphens.

Personally, I think character count is a much better and more consistent metric than word count. My main semi-professional writing experience is with essays for publications anyway, where character count is the norm. At least around here.

In any case, on the single document level this doesn't matter as much. I'm considering adding some text analysis tools on the compiled manuscript text though, where processing time is less of a concern and there is a user interface where the user can be allowed to make some decisions about how things are counted. A tricky one her is of course syllable counting. A topic I've looked into before.

**Reply by peter88213** [reply]
Since novelWriter can export the ODT format anyway, Libre/OpenOffice can already take a lot of work off your hands, since they have a sophisticated word counting algorithm as far as I can tell.
Traditionally, it's still the number of pages that counts in Germany, and that ultimately also essentially determines the printing costs.
Therefore, my yWriter to ODT exporter generates standard pages with 10-pitch fixed width font and 30 lines with maximum 60 characters. Then for Openoffice I have a ["Style Switcher" extension](https://peter88213.github.io/StyleSwitcher/) that lets you quickly switch between document styles if you want to see something different on the screen or in the printout.
